Update docs (#38)
brighthive/data-resource-generator: docs/source/architectural_design.rst

Architectural design
====================

Overview
--------

Given a configuration file (Data Resource Schema) VIA a REST API the Data Resource Generator (DRG) will dynamically generate:

#. Database tables with relationships
    * SQLAlchemy Object Relation Mappings (ORM)
    * Generate and run database migrations
#. A RESTful API
    * Running Flask application
    * Open API Specification 3.0

> 

Reviewer: I think lines 7-14 should appear in "Getting Started," since they provide a clear, enumerated description of what the DRG actually does. "Architectural Design" has more to do with "how this thing is built" and less to do with "what this thing can do."

Reviewer: Actually....the material on lines 24-27 can be moved to the "Get Started" page. I'd also suggest re-writing them:

> The DRG comes to life when a user or client sends a Data Resource Schema, embedded inside a Data Resource Generation Payload, to the DRG generation route. Upon the receipt of a Data Resource Generation Payload, the application then generates:
>
> * relevant ORM models based on your table schema document
> * relevant REST routes based on your swagger document

> The application is built with Python and uses Flask.

Reviewer: I suggest removing passive constructions when you can. This seems like a good section to use "we" and "BrightHive." I suggest condensing lines 16-22 into a clear statement of dependencies, something like:

> BrightHive built the DRG with Python and several other libraries. These include:
>
> **Table Schema and Swagger.** These open data specifications provide the backbone of a Data Resource Schema, which embeds table schema, descriptors, a swagger specification, and metadata. 
> **Flask.** The DRG application itself is a Flask API. The API has a generation route that holds a reference to the Flask object.

We leverage two open data specifications, Table Schema and Swagger. These are core to the project. By embedding table schema descriptors and a swagger specification along with some metadata you have a Data Resource Schema.

The Data Resource Schema is sent embedded inside a Data Resource Generation Payload to the DRG generation route. This triggers the generation of Data Resources.

The application will stand up as a Flask API. This API has a generation route that holds a reference to the Flask object.

Upon a Data Resource Generation Payload being sent, the application then generates:

* relevant ORM models based on your table schema document
* relevant REST routes based on your swagger document

Application Paradigm
--------------------

> "Declarative setup - imperative changes."

Reviewer: Is this a quote from somewhere? If yes, provide a citation. If not, then remove the quotes and use italic font for emphasis.

Upon initialization of your database and API the DRG uses a declarative method. If you wish to introduce changes to that declarative baseline, you are expected to do it imperatively.

> Concretely what this means is you will use a configuration file that will generate the required "stuff" to create your database and API. If you want to modify that initial configuration you will have to do so manually. The DRG, as a matter of principal, does not support migrations and modifications as such.

Reviewer: ```suggestion
What does this mean? You use a configuration file to generate the required "stuff" to create a database and API. If you want to modify the initial configuration, then you have to do so manually. The DRG, as a matter of principal, does not support migrations and modifications as such.
```

The DRG, on startup, deterministically expects a specific state of your database tables based on the provided configuration file. Therefore, you will need to update the state of your database to match the state present in your new configuration file.

Generating ORM and Tables
-------------------------

> By leveraging tooling around the table schema specification we produce a SQLAlchemy ORM. This ORM is used to generate and run the proper DDL commands on the connected database to produce and save the database tables.

Reviewer: ```suggestion
The DRG produces a SQLAlchemy ORM. This ORM is used to generate and run the proper DDL commands on the connected database to produce and save the database tables.
```

This process looks like this:

> Database Table Definition Document -> SQLAlchemy ORM -> Database Tables

Reviewer: Hmmmm....can you flesh this out a little?
